packages/vm/src/context.rs

```diff
@@ -135,13 +166,54 @@ pub(crate) fn move_into_context<S: Storage, Q: Querier>(target: &mut Ctx, storag
     b.querier = Some(querier);
 }
 
+pub fn get_gas_state<S: Storage, Q: Querier>(ctx: &mut Ctx) -> &mut GasState {
+    &mut get_context_data::<S, Q>(ctx).gas_state
+}
+
+#[cfg(feature = "default-singlepass")]
+pub fn try_consume_gas<S: Storage, Q: Querier>(ctx: &mut Ctx, used_gas: u64) -> VmResult<()> {
+    use crate::backends::{get_gas_left, set_gas_limit};
+    use crate::VmError;
+
+    let ctx_data = get_context_data::<S, Q>(ctx);
+    if let Some(mut instance_ptr) = ctx_data.wasmer_instance {
+        let instance = unsafe { instance_ptr.as_mut() };
+        let gas_state = &mut ctx_data.gas_state;
+
+        let wasmer_used_gas = gas_state.gas_limit.saturating_sub(get_gas_left(instance));
+
+        gas_state.use_gas(used_gas);
+        // These lines reduce the amount of gas available to wasmer
+        // so it can not consume gas that was consumed externally.
+        let new_limit = gas_state
```

Comment:
I would make this a method on gas_state, as this logic makes more sense there.
Something like:

```rust
pub fn wasmer_limit(&self, wasmer_used: u64) -> u64 {
  // same code
}
```

Then this looks cleaner:

```rust
gas_state.use_gas(used_gas);
let new_limit = gas_state.wasmer_limit(wasmer_used);
set_gas_limit(instance, new_limit);
```

anyway, just personal preference

Reply:
Good idea, i like it!
I'll name it "gas_left" though, and document why it needs the argument.

Reply:
done in 746f709